File: 27.RemoveElement/remove.py
# coding: utf-8 -*-

class Solution(object):
    # 将非目标值放到往前数目标值个数的位置
    def removeElement(self, nums, val):
        """
        :type nums: List[int]
        :type val: int
        :rtype: int
        """
        nTargets = 0
        for i in range(len(nums)):
            if nums[i] == val:
                nTargets += 1
            else:
                if nTargets != 0:
                    nums[i - nTargets] = nums[i]

        return len(nums) - nTargets

if __name__ == '__main__':
    nums = [2,3,2,2,3,4,3,1,1,6]
    print Solution().removeElement(nums, 3)

File: 27.RemoveElement/remove.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int nTargets = 0;
        for (unsigned int i = 0; i < nums.size(); i++) {
            if (nums[i] == val) {
                nTargets++;
            }
            else {
                if (nTargets) {
                    nums[i - nTargets] = nums[i];
                }
            }
        }
        
        return nums.size() - nTargets;
    }
};

int main(int argc, char const *argv[]) {
    int numsArr[] = {3,2,2,3};
    vector<int> nums(numsArr, numsArr + 4);
    Solution s = Solution();
    cout << s.removeElement(nums, 3) << endl;
    return 0;
}
